Neos.Fusion/Classes/Core/Cache/ParserCacheFlusher.php

```diff
@@ -50,10 +51,21 @@ public function flushPartialCacheOnFileChanges($fileMonitorIdentifier, array $ch
 
         $identifiersToFlush = [];
         foreach ($changedFiles as $changedFile => $status) {
-            // flow already returns absolute file paths from the file monitor, so we don't have to call realpath.
-            // attempting to use realpath can even result in an error as the file might be removed and thus no realpath can be resolved via file system.
-            // https://github.com/neos/neos-development-collection/pull/4509
-            $identifiersToFlush[] = $this->getCacheIdentifierForAbsoluteUnixStyleFilePathWithoutDirectoryTraversal($changedFile);
+            // flow returns linux style file paths without directory traversal from the file monitor.
+            // As discovered via https://github.com/neos/neos-development-collection/issues/4915 the paths will point to symlinks instead of the actual file.
+            // Thus, we still need to invoke `realpath` as the cache invalidation otherwise would not work (due to a different hash)
+            // But attempting to use realpath on removed/moved files fails because it surely cannot be resolved via file system.
+            if ($status === ChangeDetectionStrategyInterface::STATUS_DELETED) {
+                // Ignoring removed files means we cannot flush removed files, but this is a compromise for now.
+                // See https://github.com/neos/neos-development-collection/issues/4415 as reminder that flushing is disabled for deleted files
+                continue;
+            }
+            $fusionFileRealPath = realpath($changedFile);
+            if ($fusionFileRealPath === false) {
+                // should not happen as we ignored deleted files beforehand.
+                throw new \RuntimeException("Couldn't resolve realpath for: '$changedFile'", 1709122619);
+            }
+            $identifiersToFlush[] = $this->getCacheIdentifierForAbsoluteUnixStyleFilePathWithoutDirectoryTraversal($fusionFileRealPath);
```

Comment:
before https://github.com/neos/neos-development-collection/pull/4838 the `realpath` call would always happen inside this `ParserCacheIdentifierTrait` method. The "revert" ist that we call `realpath` again, but explicitly above, so its more obvious who and what and why.